File: frontend/src/types/invitation.ts
/**
 * Invitation related types
 */

import { BaseEntity, PaginationParams, BaseFilter } from './common';
import { User } from './auth';
import { Event } from './event';

// Invitation status
export type InvitationStatus = 'sent' | 'accepted' | 'expired' | 'cancelled';

// Invitation model
export interface Invitation extends BaseEntity {
  event_id: string;
  event: Event;
  email: string;
  token: string;
  status: InvitationStatus;
  expires_at: string;
  prefilled_name?: string;
  prefilled_phone?: string;
  prefilled_notes?: string;
  sent_at: string;
  accepted_at?: string;
  expired_at?: string;
}

// Invitation list item (simplified for list views)
export interface InvitationListItem {
  id: string;
  email: string;
  token: string;
  status: InvitationStatus;
  expires_at: string;
  prefilled_name?: string;
  sent_at: string;
  accepted_at?: string;
  created_at: string;
}

// Invitation request/response types
export interface CreateInvitationRequest {
  email: string;
  expires_in_days: number; // Required, 1-90 days
  prefilled_name?: string;
  prefilled_phone?: string;
  prefilled_notes?: string;
}

export interface UpdateInvitationRequest {
  expires_in_days?: number;
  prefilled_name?: string;
  prefilled_phone?: string;
  prefilled_notes?: string;
}

export type InvitationResponse = Invitation;

export interface InvitationsResponse {
  invitations: Invitation[];
  total: number;
}

// Invitation acceptance
export interface AcceptInvitationRequest {
  rsvp_status: 'pending' | 'accept' | 'decline' | 'maybe';
  notes?: string;
}

export interface AcceptInvitationResponse {
  guest: Guest;
  message: string;
}

// Invitation filters
export interface InvitationFilters extends BaseFilter, PaginationParams {
  event_id?: string;
  status?: InvitationStatus;
  email?: string;
  expires_after?: string;
  expires_before?: string;
  sent_after?: string;
  sent_before?: string;
}

// Invitation statistics
export interface InvitationStats {
  total_invitations: number;
  status_stats: {
    sent: number;
    accepted: number;
    expired: number;
    cancelled: number;
  };
  acceptance_rate: number;
  pending_invitations: number;
  expired_invitations: number;
}

// Bulk invitation types - imported from bulk-operations

// Resend invitation
export interface ResendInvitationRequest {
  invitation_id: string;
}

export interface ResendInvitationResponse {
  invitation: Invitation;
  message: string;
}

// Import types from guest module
import { Guest, PlusOne, PlusOneRequest } from './guest';
import { BulkInvitationRequest, BulkInvitationResult } from './bulk-operations';

File: backend/internal/database/models/invitation.go
package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

// InviteStatus represents the current status of an invitation
type InviteStatus string

const (
	InviteStatusSent      InviteStatus = "sent"
	InviteStatusAccepted  InviteStatus = "accepted"
	InviteStatusExpired   InviteStatus = "expired"
	InviteStatusCancelled InviteStatus = "cancelled"
)

// Invitation represents an invitation to an event
type Invitation struct {
	ID      uuid.UUID `json:"id" gorm:"type:uuid;primary_key;default:gen_random_uuid()"`
	EventID uuid.UUID `json:"event_id" gorm:"type:uuid;not null"`
	Event   Event     `json:"event" gorm:"foreignKey:EventID"`

	// Invitation Details
	Email     string       `json:"email" gorm:"not null;index"`
	Token     string       `json:"token" gorm:"uniqueIndex;not null"`
	Status    InviteStatus `json:"status" gorm:"not null;default:'sent'"`
	ExpiresAt time.Time    `json:"expires_at" gorm:"not null"`

	// Pre-filled Information (optional)
	PrefilledName  *string `json:"prefilled_name"`
	PrefilledPhone *string `json:"prefilled_phone"`
	PrefilledNotes *string `json:"prefilled_notes"`

	// Tracking
	SentAt     time.Time  `json:"sent_at"`
	AcceptedAt *time.Time `json:"accepted_at"`
	ExpiredAt  *time.Time `json:"expired_at"`

	// Timestamps
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `json:"-" gorm:"index"`
}

// CreateInvitationRequest represents the request to create an invitation
type CreateInvitationRequest struct {
	Email          string  `json:"email" binding:"required,email" example:"guest@example.com"`
	PrefilledName  *string `json:"prefilled_name" example:"John Doe"`
	PrefilledPhone *string `json:"prefilled_phone" example:"+1234567890"`
	PrefilledNotes *string `json:"prefilled_notes" example:"Vegetarian"`
	ExpiresInDays  int     `json:"expires_in_days" binding:"required,min=1,max=90" example:"30"`
}

// UpdateInvitationRequest represents the request to update an invitation
type UpdateInvitationRequest struct {
	PrefilledName  *string `json:"prefilled_name" example:"John Doe"`
	PrefilledPhone *string `json:"prefilled_phone" example:"+1234567890"`
	PrefilledNotes *string `json:"prefilled_notes" example:"Vegetarian"`
	ExpiresInDays  *int    `json:"expires_in_days" binding:"omitempty,min=1,max=90" example:"30"`
}

// InvitationResponse represents the invitation data in API responses
type InvitationResponse struct {
	ID             uuid.UUID    `json:"id"`
	EventID        uuid.UUID    `json:"event_id"`
	Email          string       `json:"email"`
	Status         InviteStatus `json:"status"`
	ExpiresAt      time.Time    `json:"expires_at"`
	PrefilledName  *string      `json:"prefilled_name"`
	PrefilledPhone *string      `json:"prefilled_phone"`
	PrefilledNotes *string      `json:"prefilled_notes"`
	SentAt         time.Time    `json:"sent_at"`
	AcceptedAt     *time.Time   `json:"accepted_at"`
	ExpiredAt      *time.Time   `json:"expired_at"`
	CreatedAt      time.Time    `json:"created_at"`
	UpdatedAt      time.Time    `json:"updated_at"`
}

// InvitationListItem represents invitation data in list responses
type InvitationListItem struct {
	ID            uuid.UUID    `json:"id"`
	Email         string       `json:"email"`
	Token         string       `json:"token"`
	Status        InviteStatus `json:"status"`
	ExpiresAt     time.Time    `json:"expires_at"`
	PrefilledName *string      `json:"prefilled_name"`
	SentAt        time.Time    `json:"sent_at"`
	AcceptedAt    *time.Time   `json:"accepted_at"`
	CreatedAt     time.Time    `json:"created_at"`
}

// AcceptInvitationRequest represents the request to accept an invitation
type AcceptInvitationRequest struct {
	RSVPStatus RSVPStatus `json:"rsvp_status" binding:"required,oneof=pending accept decline maybe" example:"accept"`
	Notes      *string    `json:"notes" example:"Vegetarian, allergic to nuts"`
}

// AcceptInvitationResponse represents the response after accepting an invitation
type AcceptInvitationResponse struct {
	Guest   Guest  `json:"guest"`
	Message string `json:"message" example:"Successfully accepted invitation"`
}
